README.md
# image-difference

> Create image differential between two images

[![npm package](https://img.shields.io/npm/v/image-difference.svg)](https://www.npmjs.org/package/image-difference)
[![Build Status](https://travis-ci.org/argos-ci/image-difference.svg?branch=master)](https://travis-ci.org/argos-ci/image-difference)

[![PeerDependencies](https://img.shields.io/david/peer/argos-ci/image-difference.svg)](https://david-dm.org/argos-ci/image-difference#info=peerDependencies&view=list)
[![Dependencies](https://img.shields.io/david/argos-ci/image-difference.svg)](https://david-dm.org/argos-ci/image-difference)
[![DevDependencies](https://img.shields.io/david/dev/argos-ci/image-difference.svg)](https://david-dm.org/argos-ci/image-difference#info=devDependencies&view=list)

## The problem solved

![difference](example/example.png)

This is a fork of [image-diff](https://github.com/uber-archive/image-diff) that is no longer maintained.
This was created as part of a [visual regression](http://www.youtube.com/watch?v=1wHr-O6gEfc) project and now supported as part of [Argos-CI](https://www.argos-ci.com/).

## Installation

`image-difference` depends on [ImageMagick](http://www.imagemagick.org/script/index.php) Please install this before continuing.

```sh
npm install --save image-difference
```

## API

### `imageDifference(options) => Promise`

Create image differential between two images

#### Arguments

1. `options` (*Object*)
    - `options.actualFilename` (*String*): Path to actual image file. **must** exist.
    - `options.expectedFilename` (*String*): Path to expected image file. **must** exist.
    - `options.diffFilename` (*String*): Optional path to output differential image.
    - `options.metric` (*String*): Optional metric used for the computation of the output.
        - [documentation](http://legacy.imagemagick.org/script/command-line-options.php#metric)
        - [guide](http://www.imagemagick.org/Usage/compare/)
        - [implementation](https://github.com/ImageMagick/ImageMagick/blob/master/MagickCore/compare.c)

#### Returns

`Promise`: Return the difference object with a `total` and `percentage` key.

#### [Example](https://github.com/argos-ci/image-difference/tree/master/example)

```js
import imageDifference from '../src/imageDifference'

imageDifference(
  {
    actualFilename: `${__dirname}/images/hello-world.png`,
    expectedFilename: `${__dirname}/images/hello.png`,
    diffFilename: `${__dirname}/images/hello-diff.png`,
  },
  console.log
)
```

## CLI Usage

```sh
Usage:  [options] <actual-filename> <expected-filename> [diff-filename]

Create image differential between two images

Options:

  -h, --help     output usage information
  -V, --version  output the version number
```


## Benchmark

We use `imagemagick1` as the default implmentation

```sh
image-diff diff x 1.11 ops/sec ±1.21% (25 runs sampled)
image-diff same x 1.07 ops/sec ±0.80% (25 runs sampled)

image-difference imagemagick1 diff x 4.43 ops/sec ±0.40% (41 runs sampled)
image-difference imagemagick1 same x 4.45 ops/sec ±0.49% (41 runs sampled)

image-difference imagemagick2 diff x 2.96 ops/sec ±0.49% (34 runs sampled)
image-difference imagemagick2 same x 2.94 ops/sec ±0.87% (34 runs sampled)

image-difference graphicsmagick diff x 2.32 ops/sec ±0.60% (31 runs sampled)
image-difference graphicsmagick same x 2.30 ops/sec ±0.72% (31 runs sampled)

pixelmatch diff x 0.79 ops/sec ±0.94% (23 runs sampled)
pixelmatch same x 3.58 ops/sec ±2.07% (37 runs sampled)

resemble diff x 0.59 ops/sec ±4.76% (22 runs sampled)
resemble same x 0.94 ops/sec ±0.85% (24 runs sampled)

looksSame diff x 2.09 ops/sec ±3.08% (30 runs sampled)
looksSame same x 1.08 ops/sec ±3.76% (25 runs sampled)
```

## Contributing

```sh
brew install imagemagick@6 graphicsmagick
yarn
yarn test:watch
```

## License

MIT


CHANGELOG.md
All notable changes are described on the [Releases](https://github.com/argos-ci/image-diff/releases) page.


test/e2e.test.js
import rimraf from 'rimraf'
import getPixels from 'get-pixels'
import imageDifference from '../src/imageDifference'

function getPixelsAsync(filename) {
  return new Promise((accept, reject) => {
    getPixels(filename, (err, pixels) => {
      if (err) {
        reject(err)
        return
      }

      accept(pixels)
    })
  })
}

// Clean up actual-files
rimraf.sync(`${__dirname}/actual-files`)

describe('e2e', () => {
  it('diffing different images', async () => {
    const difference = await imageDifference({
      actualFilename: `${__dirname}/test-files/checkerboard.png`,
      expectedFilename: `${__dirname}/test-files/white.png`,
      diffFilename: `${__dirname}/actual-files/different.png`,
    })
    const actual = await getPixelsAsync(`${__dirname}/actual-files/different.png`)
    const expected = await getPixelsAsync(`${__dirname}/expected-files/different.png`)

    expect(difference).toMatchSnapshot()
    expect(actual).toEqual(expected)
  })

  it('diffing the same image', async () => {
    const difference = await imageDifference({
      actualFilename: `${__dirname}/test-files/checkerboard.png`,
      expectedFilename: `${__dirname}/test-files/checkerboard.png`,
      diffFilename: `${__dirname}/actual-files/same.png`,
    })
    const actual = await getPixelsAsync(`${__dirname}/actual-files/same.png`)
    const expected = await getPixelsAsync(`${__dirname}/expected-files/same.png`)

    expect(difference).toEqual({
      value: 0,
      width: 10,
      height: 10,
    })
    expect(actual).toEqual(expected)
  })

  it('diffing different sizes images', async () => {
    const difference = await imageDifference({
      actualFilename: `${__dirname}/test-files/checkerboard-excess.png`,
      expectedFilename: `${__dirname}/test-files/checkerboard.png`,
      diffFilename: `${__dirname}/actual-files/different-size.png`,
    })
    const actual = await getPixelsAsync(`${__dirname}/actual-files/different-size.png`)
    const expected = await getPixelsAsync(`${__dirname}/expected-files/different-size.png`)

    expect(difference).toMatchSnapshot()
    expect(actual).toEqual(expected)
  })

  it('diffing different images without an output image', async () => {
    const difference = await imageDifference({
      actualFilename: `${__dirname}/test-files/checkerboard.png`,
      expectedFilename: `${__dirname}/test-files/white.png`,
    })

    expect(difference).toMatchSnapshot()
  })

  it('diffing images which cannot scale into each other', async () => {
    const difference = await imageDifference({
      actualFilename: `${__dirname}/test-files/horizontal.png`,
      expectedFilename: `${__dirname}/test-files/vertical.png`,
      diffFilename: `${__dirname}/actual-files/horizontal-vertical.png`,
    })
    const actual = await getPixelsAsync(`${__dirname}/actual-files/horizontal-vertical.png`)
    const expected = await getPixelsAsync(`${__dirname}/expected-files/horizontal-vertical.png`)

    expect(difference).toMatchSnapshot()
    expect(actual).toEqual(expected)
  })

  it('diffing the same image where 1 has a transparent background', async () => {
    const difference = await imageDifference({
      actualFilename: `${__dirname}/test-files/checkerboard-transparent.png`,
      expectedFilename: `${__dirname}/test-files/checkerboard.png`,
      diffFilename: `${__dirname}/actual-files/different-transparent.png`,
    })
    expect(difference).toMatchSnapshot()
  })
})


test/benchmark/resemble.js
/* eslint-disable no-console, no-use-before-define */

import rimraf from 'rimraf'
import fs from 'fs'
import Benchmark from 'benchmark'
import resemble from 'resemblejs-node'

// Clean up actual-files
rimraf.sync(`${__dirname}/../actual-files`)

const suite = new Benchmark.Suite()

suite
  .add('resemble diff', {
    minSamples: 20,
    defer: true,
    fn: deferred => {
      const diff = resemble(`${__dirname}/base.png`).compareTo(`${__dirname}/compare.png`)
      diff.onComplete(diffResult => {
        if (diffResult.rawMisMatchPercentage > 0) {
          diffResult
            .getDiffImage()
            .pack()
            .pipe(fs.createWriteStream('diff.png'))
            .on('finish', () => {
              deferred.resolve()
            })
        } else {
          deferred.resolve()
        }
      })
    },
  })
  .add('resemble same', {
    minSamples: 20,
    defer: true,
    fn: deferred => {
      const diff = resemble(`${__dirname}/base.png`).compareTo(`${__dirname}/base.png`)
      diff.onComplete(diffResult => {
        if (diffResult.rawMisMatchPercentage > 0) {
          diffResult
            .getDiffImage()
            .pack()
            .pipe(fs.createWriteStream('diff.png'))
            .on('finish', () => {
              deferred.resolve()
            })
        } else {
          deferred.resolve()
        }
      })
    },
  })
  .on('cycle', event => {
    console.log(String(event.target))
  })
  .run({ async: true })


src/imageDifference.test.js
import { handleRaw } from './imageDifference'

describe('handleRaw', () => {
  describe('An image-difference output with no difference', () => {
    it('extracts no difference', () => {
      const difference = handleRaw(
        {
          raw: 'all: 0 (0)',
        },
        {
          metric: 'RMSE',
        }
      )
      expect(difference.value1).toBe(0)
      expect(difference.value2).toBe(0)
    })
  })

  describe('An image-difference output with a significant difference', () => {
    it('extracts the significant difference', () => {
      const difference = handleRaw(
        {
          raw: 'all: 40131.8 (0.612372)',
        },
        {
          metric: 'RMSE',
        }
      )
      expect(difference.value1).toBe(40131.8)
      expect(difference.value2).toBe(0.612372)
    })
  })

  describe('An image-difference output with a fractional difference', () => {
    it('extracts the fractional difference', () => {
      const difference = handleRaw(
        { raw: 'all: 0.460961 (7.03381e-06)' },
        {
          metric: 'RMSE',
        }
      )
      expect(difference.value1).toBe(0.460961)
      expect(difference.value2).toBe(7.03381e-6)
    })
  })

  describe('An image-difference output with a super fractional difference', () => {
    it('extracts the fractional difference', () => {
      const difference = handleRaw(
        { raw: 'all: 7.03381e-06 (7.03381e-06)' },
        {
          metric: 'RMSE',
        }
      )
      expect(difference.value1).toBe(7.03381e-6)
      expect(difference.value2).toBe(7.03381e-6)
    })
  })
})


.eslintrc.js
module.exports = {
  root: true, // So parent files don't get applied
  env: {
    es6: true,
    browser: true,
    node: true,
    jest: true,
  },
  extends: ['airbnb', 'plugin:import/recommended', 'plugin:jest/recommended'],
  parser: 'babel-eslint',
  parserOptions: {
    ecmaVersion: 7,
    sourceType: 'module',
  },
  plugins: ['babel', 'import', 'prettier', 'jest'],
  settings: {
    'import/resolver': {
      node: {
        // As configured in webpack
        moduleDirectory: ['node_modules', 'src'],
      },
    },
  },
  rules: {
    'arrow-body-style': 'off', // Not our taste?
    'max-len': 'off', // Not our taste?
    'arrow-parens': 'off', // Incompatible with prettier
    'no-confusing-arrow': 'off', // Incompatible with prettier
    indent: 'off', // Incompatible with prettier
    'space-before-function-paren': 'off', // Incompatible with prettier
    semi: ['error', 'never'],
    'consistent-this': ['error', 'self'],
    'no-console': 'error', // airbnb is using warn
    'no-alert': 'error', // airbnb is using warn
    'object-curly-spacing': 'off', // use babel plugin rule
    'no-restricted-properties': 'off', // To remove once react-docgen support ** operator.
    'no-mixed-operators': 'off', // allow a + b * c instead of a + (b * c), prettier conflict
    'comma-dangle': [
      'error',
      {
        arrays: 'always-multiline',
        objects: 'always-multiline',
        imports: 'always-multiline',
        exports: 'always-multiline',
        functions: 'never',
      },
    ],

    'babel/object-curly-spacing': ['error', 'always'],

    'import/unambiguous': 'off',
    'import/no-unresolved': 'off',
    'import/no-named-as-default': 'off',
    'import/extensions': 'off',
    'import/no-extraneous-dependencies': 'off',
    'import/prefer-default-export': 'off',

    'prettier/prettier': [
      'error',
      {
        singleQuote: true,
        printWidth: 100,
        trailingComma: 'es5',
        semi: false,
      },
    ],
  },
}


src/command.js
/* eslint-disable no-console */

import program from 'commander'
import chalk from 'chalk'
import pkg from '../package.json'
import imageDifference from './imageDifference'
import { displayError, displaySuccess } from './display'

program
  .version(pkg.version)
  .description('Create image differential between two images')
  .usage('[options] <actual-filename> <expected-filename> [diff-filename]')
  .action(async (actualFilename, expectedFilename, diffFilenameArgs, command) => {
    let options = command
    let diffFilename = diffFilenameArgs

    // If there is no program, then assume diffFilename was left out
    if (command === undefined) {
      options = diffFilenameArgs
      diffFilename = undefined
    }

    let difference

    try {
      difference = await imageDifference({
        actualFilename,
        diffFilename,
        expectedFilename,
        ...options,
      })
    } catch (error) {
      displayError('Sorry an error happened:')

      console.error(chalk.bold.red(error.stack))
      process.exit(1)
    }

    if (difference.value === 0) {
      displaySuccess('No difference')
    } else {
      displayError(
        `A difference was detected (value: ${difference.value}, percentage: ${difference.value /
          (difference.width * difference.height)})`
      )
      process.exit(1)
    }
  })

if (!process.argv.slice(3).length) {
  program.outputHelp()
} else {
  program.parse(process.argv)
}


src/imageDifference.js
import fs from 'fs'
import path from 'path'
import gm from 'gm'
import spawn from 'cross-spawn'
import mkdirp from 'mkdirp'
import tmp from 'tmp'

const gmMagick = gm.subClass({ imageMagick: true })

function transparent(filename, options) {
  if (!options.width || !options.height) {
    throw new Error('Wrong options provided to transparent()')
  }

  const gmImage = gmMagick(filename)
  gmImage.background('transparent') // Fill in new space with white background
  gmImage.gravity('NorthWest') // Anchor image to upper-left
  gmImage.extent(options.width, options.height) // Specify new image size

  return gmImage
}

function getImageSize(filename) {
  return new Promise((accept, reject) => {
    fs.stat(filename, err => {
      if (err) {
        reject(err)
        return
      }
      gmMagick(filename).size((err2, value) => {
        if (err2) {
          reject(err2)
          return
        }
        accept(value)
      })
    })
  })
}

export function handleRaw(difference, options) {
  const { raw, ...other } = difference

  // Only process once
  if (typeof raw === 'number') {
    return difference
  }

  let resultInfo

  switch (options.metric) {
    case 'PAE':
    case 'MSE':
    case 'RMSE':
    case 'MAE':
      // Find variant between 'all: 0 (0)', 'all: 40131.8 (0.612372)', or 'all: 0.460961 (7.03381e-06)'
      // According to http://www.imagemagick.org/discourse-server/viewtopic.php?f=1&t=17284
      // These values are the total root mean square (MSE) pixel difference
      // across all pixels and its percentage
      resultInfo = raw.match(/all: (\d+(?:\.\d+)?(?:[Ee]-?\d+)?) \((\d+(?:\.\d+)?(?:[Ee]-?\d+)?)\)/)
      if (!resultInfo) {
        throw new Error(`Expected raw to contain 'all' but received "${raw}"`)
      }

      return {
        ...other,
        value1: parseFloat(resultInfo[1], 10),
        value2: parseFloat(resultInfo[2], 10),
      }

    case 'PHASH':
    case 'PSNR':
    case 'AE':
    case 'NCC':
    case 'FUZZ':
      resultInfo = raw.match(/all: (\d+(\.\d+)?)/)
      if (!resultInfo) {
        throw new Error(`Expected raw to contain 'all' but received "${raw}"`)
      }
      return {
        ...other,
        value: parseFloat(resultInfo[1], 10),
      }

    default:
      throw new Error('Unknown metric')
  }
}

function resizeImage(filename, options) {
  if (!options.width || !options.height) {
    throw new Error('Wrong options provided to resizeImage()')
  }

  return new Promise((accept, reject) => {
    // Get a temporary filepath
    tmp.tmpName({ postfix: '.png' }, (err, tmpFilename) => {
      // If there was an error, callback
      if (err) {
        reject(err)
        return
      }

      transparent(filename, {
        width: options.width,
        height: options.height,
      }).write(tmpFilename, err2 => {
        if (err2) {
          reject(err2)
          return
        }

        accept(tmpFilename)
      })
    })
  })
}

function createDifference(options) {
  const {
    actualFilename,
    expectedFilename,
    diffFilename,
    implementation,
    metric,
    highlightColor,
    lowlightColor,
    fuzz,
  } = options

  const diffArgs = [
    '-verbose',
    '-highlight-color',
    highlightColor,
    '-lowlight-color',
    lowlightColor,
    // http://legacy.imagemagick.org/script/command-line-options.php#metric
    // http://www.imagemagick.org/Usage/compare/
    // https://github.com/ImageMagick/ImageMagick/blob/master/MagickCore/compare.c
    '-metric',
    metric,
  ]
    .concat(fuzz ? ['-fuzz', fuzz] : [])
    // Paths to actual, expected, and diff images
    .concat([
      actualFilename,
      expectedFilename,
      // If there is no output image, then output to `stdout` (which is ignored)
      diffFilename || '-',
    ])

  return new Promise((accept, reject) => {
    switch (implementation) {
      case 'imagemagick1': {
        // http://www.imagemagick.org/script/compare.php
        const proc = spawn('compare', diffArgs)
        let stdout = ''
        let stderr = ''
        proc.stdout.on('data', data => {
          stdout += data
        })
        proc.stderr.on('data', data => {
          stderr += data
        })
        proc.on('close', code => {
          const isImageMagick = true
          // ImageMagick returns err code 2 if err, 0 if similar, 1 if dissimilar
          if (isImageMagick) {
            if (code === 0 || code === 1) {
              stdout = stderr
            } else {
              reject(stderr)
              return
            }
          } else if (code !== 0) {
            reject(stderr)
            return
          }
          accept(stdout)
        })
        break
      }

      case 'imagemagick2':
        // http://www.imagemagick.org/script/compare.php
        // https://github.com/aheckmann/gm/blob/master/lib/compare.js
        gmMagick().compare(
          actualFilename,
          expectedFilename,
          {
            file: diffFilename,
          },
          (err, Boolean, equality) => {
            if (err) {
              reject(err)
              return
            }

            accept(equality)
          }
        )
        break

      case 'graphicsmagick': {
        // http://www.graphicsmagick.org/GraphicsMagick.html
        // https://github.com/aheckmann/gm/blob/master/lib/compare.js
        gm.compare(
          actualFilename,
          expectedFilename,
          {
            highlightColor: 'RED',
            highlightStyle: 'Assign',
            file: diffFilename,
          },
          (err, Boolean, equality) => {
            if (err) {
              reject(err)
              return
            }

            accept(equality)
          }
        )
        break
      }

      default:
        throw new Error('Unknow implementation')
    }
  })
}

export async function rawDifference(options) {
  const { actualFilename, expectedFilename, diffFilename, ...other } = options

  await Promise.all(
    [actualFilename, expectedFilename].map(
      filename =>
        new Promise((accept, reject) => {
          fs.exists(filename, actualExists => {
            if (actualExists) {
              accept()
              return
            }

            reject(
              new Error(
                `\`image-difference\` expected "${filename}" filename to exist but it didn't`
              )
            )
          })
        })
    )
  )

  const sizes = await Promise.all([getImageSize(actualFilename), getImageSize(expectedFilename)])

  // Find the maximum dimensions
  const actualSize = sizes[0]
  const expectedSize = sizes[1]
  const maxWidth = Math.max(actualSize.width, expectedSize.width)
  const maxHeight = Math.max(actualSize.height, expectedSize.height)

  let actualTmpFilename
  let expectedTmpFilename

  if (actualSize.width !== maxWidth || actualSize.height !== maxHeight) {
    actualTmpFilename = await resizeImage(actualFilename, {
      width: maxWidth,
      height: maxHeight,
    })
  }

  if (expectedSize.width !== maxWidth || expectedSize.height !== maxHeight) {
    expectedTmpFilename = await resizeImage(expectedFilename, {
      width: maxWidth,
      height: maxHeight,
    })
  }

  if (diffFilename) {
    await new Promise((accept, reject) => {
      mkdirp(path.dirname(diffFilename), err => {
        if (err) {
          reject(err)
          return
        }

        accept()
      })
    })
  }

  const raw = await createDifference({
    actualFilename: actualTmpFilename || actualFilename,
    expectedFilename: expectedTmpFilename || expectedFilename,
    diffFilename,
    ...other,
  })

  // Clean up the temporary files
  await Promise.all(
    [actualTmpFilename, expectedTmpFilename].filter(filename => filename).map(
      filename =>
        new Promise((accept, reject) => {
          fs.unlink(filename, err => {
            if (err) {
              reject(err)
              return
            }

            accept()
          })
        })
    )
  )

  return {
    raw,
    width: maxWidth,
    height: maxHeight,
  }
}

export default async function imageDifference(optionsWithoutDefault) {
  const {
    actualFilename,
    expectedFilename,
    implementation = 'imagemagick1',
    metric = 'AE',
    highlightColor = 'red',
    lowlightColor = 'white',
    fuzz = 0,
    ...other
  } = optionsWithoutDefault

  // Assert our options are passed in
  if (!actualFilename) {
    throw new Error('`options.actualFilename` was not passed to `image-difference`')
  }

  if (!expectedFilename) {
    throw new Error('`options.expectedFilename` was not passed to `image-difference`')
  }

  const options = {
    actualFilename,
    expectedFilename,
    implementation,
    metric,
    highlightColor,
    lowlightColor,
    fuzz,
    ...other,
  }

  const difference = await rawDifference(options)
  return handleRaw(difference, options)
}
